src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -136,18 +136,27 @@ Object* FrozenObjectSegment::TryAllocateObject(PTR_MethodTable type, size_t obje
 {
     _ASSERT(m_pStart != nullptr && FOH_SEGMENT_SIZE > 0 && m_SegmentHandle != nullptr); // Expected to be inited
     _ASSERT(IS_ALIGNED(m_pCurrent, DATA_ALIGNMENT));
+    _ASSERT(objectSize <= FOH_COMMIT_SIZE);
+    _ASSERT(m_pCurrent >= m_pStart + sizeof(ObjHeader));
 
-    uint8_t* obj = m_pCurrent;
-    if (reinterpret_cast<size_t>(m_pStart + FOH_SEGMENT_SIZE) < reinterpret_cast<size_t>(obj + objectSize))
+    const size_t spaceUsed = (size_t)(m_pCurrent - m_pStart);
+    const size_t spaceLeft = FOH_SEGMENT_SIZE - spaceUsed;
+
+    _ASSERT(spaceUsed >= sizeof(ObjHeader));
+    _ASSERT(spaceLeft >= sizeof(ObjHeader));
+
+    // Test if we have a room for the given object (including extra sizeof(ObjHeader) for next object)
+    if (spaceLeft - sizeof(ObjHeader) < objectSize)
     {
-        // Segment is full
         return nullptr;
     }
 
     // Check if we need to commit a new chunk
-    if (reinterpret_cast<size_t>(m_pStart + m_SizeCommitted) < reinterpret_cast<size_t>(obj + objectSize))
+    if (spaceUsed + objectSize + sizeof(ObjHeader) > m_SizeCommitted)
```

Comment:
this can not overflow because it's already proven to this point that there is a room for new object